File: .agents/skills/do-web-doc-resolver/scripts/routing_memory.py
"""
Per-domain routing memory for the Web Doc Resolver.
"""

from collections import defaultdict


class RoutingMemory:
    def __init__(self):
        # domain -> provider -> stats
        self.domain_stats = defaultdict(
            lambda: defaultdict(
                lambda: {
                    "success": 0,
                    "failure": 0,
                    "avg_latency_ms": 0.0,
                    "avg_quality": 0.0,
                }
            )
        )

    def record(
        self, domain: str, provider: str, success: bool, latency_ms: int, quality_score: float
    ) -> None:
        stats = self.domain_stats[domain][provider]
        total = stats["success"] + stats["failure"]
        stats["avg_latency_ms"] = ((stats["avg_latency_ms"] * total) + latency_ms) / (total + 1)
        stats["avg_quality"] = ((stats["avg_quality"] * total) + quality_score) / (total + 1)
        if success:
            stats["success"] += 1
        else:
            stats["failure"] += 1

    def rank(self, domain: str, providers: list[str]) -> list[str]:
        if domain not in self.domain_stats:
            return providers

        def provider_score(provider: str) -> tuple[float, float, float]:
            s = self.domain_stats[domain][provider]
            total = s["success"] + s["failure"]
            success_rate = (s["success"] / total) if total else 0.5
            # Rank by success rate, then quality, then (negative) latency
            return (success_rate, s["avg_quality"], -s["avg_latency_ms"])

        return sorted(providers, key=provider_score, reverse=True)

    def get_p75_latency(self, domain: str, provider: str, default: int = 2500) -> int:
        stats = self.domain_stats.get(domain, {}).get(provider)
        if not stats or stats["avg_latency_ms"] == 0:
            return default
        # Heuristic: p75 is often ~1.5x average for tail latency
        return int(stats["avg_latency_ms"] * 1.5)

File: cli/src/routing_memory.rs
use chrono::{DateTime, Utc};
use std::collections::HashMap;

#[derive(Debug, Clone, Default)]
pub struct ProviderStats {
    pub success: usize,
    pub failure: usize,
    pub avg_latency_ms: f32,
    pub avg_quality: f32,
    pub last_attempted: Option<DateTime<Utc>>,
}

#[derive(Default)]
pub struct RoutingMemory {
    domain_stats: HashMap<String, HashMap<String, ProviderStats>>,
}

impl RoutingMemory {
    pub fn record(
        &mut self,
        domain: &str,
        provider: &str,
        success: bool,
        latency_ms: u64,
        quality_score: f32,
    ) {
        let providers = self.domain_stats.entry(domain.to_string()).or_default();
        let stats = providers.entry(provider.to_string()).or_default();
        let total = stats.success + stats.failure;
        let total_f = total as f32;

        stats.avg_latency_ms =
            ((stats.avg_latency_ms * total_f) + latency_ms as f32) / (total_f + 1.0);
        stats.avg_quality = ((stats.avg_quality * total_f) + quality_score) / (total_f + 1.0);
        stats.last_attempted = Some(Utc::now());

        if success {
            stats.success += 1;
        } else {
            stats.failure += 1;
        }
    }

    pub fn compute_score(&self, provider: &str, domain: &str) -> f64 {
        let Some(domain_map) = self.domain_stats.get(domain) else {
            return 0.5;
        };
        let Some(stats) = domain_map.get(provider) else {
            return 0.5;
        };

        let attempts = stats.success + stats.failure;
        if attempts == 0 {
            return 0.5;
        }

        let success_rate = stats.success as f64 / attempts as f64;

        let days_since_last = if let Some(last) = stats.last_attempted {
            let duration = Utc::now().signed_duration_since(last);
            duration.num_seconds() as f64 / 86400.0
        } else {
            0.0
        };

        let quality_factor = 0.5 + 0.5 * stats.avg_quality as f64;
        let recency_weight = (-days_since_last / 7.0).exp();
        let score = (success_rate * quality_factor * recency_weight) * 1000.0
            / (stats.avg_latency_ms as f64).max(1.0);

        tracing::debug!(
            "Provider score: domain={}, provider={}, score={:.4}, success_rate={:.2}, quality={:.2}, recency={:.2}, latency={:.1}ms",
            domain,
            provider,
            score,
            success_rate,
            stats.avg_quality,
            recency_weight,
            stats.avg_latency_ms
        );

        score
    }

    pub fn rank_providers(&self, domain: &str, providers: &[String]) -> Vec<String> {
        let mut scores: Vec<(&String, f64)> = providers
            .iter()
            .map(|p| (p, self.compute_score(p, domain)))
            .collect();

        scores.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        scores.into_iter().map(|(p, _)| p.clone()).collect()
    }
}
